trieste/acquisition/function/multi_objective.py

```diff
@@ -379,3 +384,245 @@ def _update_expected_improvement_fn(
+        :param new_optimization_step: Indicates whether this call to update_acquisition_function
+            is to start of a new optimization step, of to continue collecting batch of points
+            for the current step. Defaults to ``True``.
+        :return: The updated acquisition function.
+        """
+        tf.debugging.Assert(datasets is not None, [])
+        datasets = cast(Mapping[str, Dataset], datasets)
+        tf.debugging.Assert(datasets[self._objective_tag] is not None, [])
+        tf.debugging.assert_positive(
+            len(datasets[self._objective_tag]),
+            message=f"{self._objective_tag} dataset must be populated.",
+        )
+        tf.debugging.Assert(self._base_acquisition_function is not None, [])
+
+        if new_optimization_step:
+            self._update_base_acquisition_function(models, datasets)
+
+        if pending_points is None or len(pending_points) == 0:
+            # no penalization required if no pending_points
+            return cast(AcquisitionFunction, self._base_acquisition_function)
+
+        return self._update_penalization(function, models[self._objective_tag], pending_points)
+
+    def _update_penalization(
+        self,
+        function: AcquisitionFunction,
+        model: ProbabilisticModel,
+        pending_points: Optional[TensorType] = None,
+    ) -> AcquisitionFunction:
+        tf.debugging.assert_rank(pending_points, 2)
+
+        if self._penalized_acquisition is not None and isinstance(
+            self._penalization, hippo_penalizer
+        ):
+            # if possible, just update the penalization function variables
+            self._penalization.update(pending_points)
+            return self._penalized_acquisition
+        else:
+            # otherwise construct a new penalized acquisition function
+            self._penalization = hippo_penalizer(model, pending_points)
+
+        @tf.function
+        def penalized_acquisition(x: TensorType) -> TensorType:
+            log_acq = tf.math.log(
+                cast(AcquisitionFunction, self._base_acquisition_function)(x)
+            ) + tf.math.log(cast(PenalizationFunction, self._penalization)(x))
+            return tf.math.exp(log_acq)
+
+        self._penalized_acquisition = penalized_acquisition
+        return penalized_acquisition
+
+    def _update_base_acquisition_function(
+        self,
+        models: Mapping[str, ProbabilisticModelType],
+        datasets: Optional[Mapping[str, Dataset]] = None,
+    ) -> AcquisitionFunction:
+        if self._base_acquisition_function is None:
+            self._base_acquisition_function = self._base_builder.prepare_acquisition_function(
+                models, datasets
+            )
+        else:
+            self._base_acquisition_function = self._base_builder.update_acquisition_function(
+                self._base_acquisition_function, models, datasets
+            )
+        return self._base_acquisition_function
+
+
+class hippo_penalizer:
```

Comment:
Can this inherit from UpdateablePenalizationFunction?

Reply:
Alas! `UpdateablePenalizationFunction` makes some LP-specific references, namely

        lipschitz_constant: TensorType,
        eta: TensorType,

Reply:
ofc. Ignore me

Reply:
That would allow you to use hippo_penalizer directly with LocalPenalization. Does that make sense? If so, what's the difference between that and using HippoAcqusuisitionFunction?

Note also that the update method for UpdateablePenalizationFunction has lipschitz_constant and eta parameters that we don't need here, but we could always just ignore those.

Reply:
Great question! We never considered HIPPO for single objective use case, but perhaps we should, @henrymoss ?

The main differences now are handling of multiple models and multi-objective defaults.

Ignoring parameters is something I would like to avoid if possible. It's very confusing to have a parameter that isn't used. There are precedents in Trieste of parameters being unused and thus documented like that, so it's not the end of the world if we do it here too. But I'd rather not